src/main/java/org/folio/coursereserves/util/CRUtil.java

```diff
@@ -116,86 +116,50 @@ public static Map<String, String> getTextAcceptHeaders() {
       Map<String, String> okapiHeaders, Context context) {
-    Promise<JsonObject> promise = Promise.promise();
-    Promise<String> itemIdPromise = Promise.promise();
     String barcode;
-    if(reserve.getCopiedItem() != null) {
+    if (reserve.getCopiedItem() != null) {
       barcode = reserve.getCopiedItem().getBarcode();
     } else {
       barcode = null;
     }
     String itemId = reserve.getItemId();
-    if(itemId != null && barcode == null) {
-      itemIdPromise.complete(itemId);
-    } else if(barcode != null) {
-      lookupItemByBarcode(barcode, okapiHeaders, context)
-          .onComplete(barcodeItemLookupRes -> {
-        if(barcodeItemLookupRes.failed()) {
-          itemIdPromise.fail(barcodeItemLookupRes.cause());
-        } else {
-          if(barcodeItemLookupRes.result() != null) {
-            itemIdPromise.complete(barcodeItemLookupRes.result().getString("id"));
-          } else {
-            itemIdPromise.fail("No item found for barcode " + barcode);
-          }
+    Future<String> itemIdFuture;
+    if (itemId != null && barcode == null) {
+      itemIdFuture = Future.succeededFuture(itemId);
+    } else if (barcode != null) {
+      itemIdFuture = lookupItemByBarcode(barcode, okapiHeaders, context).map(barcodeItemLookupRes -> {
+        if (barcodeItemLookupRes == null) {
+          throw new RuntimeException("No item found for barcode " + barcode);
         }
+        return barcodeItemLookupRes.getString("id");
       });
     } else {
-      promise.fail("Must provide item id or item barcode to populate copied items");
-      return promise.future();
-    }
-    itemIdPromise.future().onComplete(itemIdRes -> {
-      if(itemIdRes.failed()) {
-        logger.error("Failed to get item id {}", itemIdRes.cause().getMessage());
-        promise.fail(itemIdRes.cause());
-      } else {
-        reserve.setItemId(itemIdRes.result());
-        String retrievedItemId = itemIdRes.result();
-        logger.info("Looking up information for item {} from inventory module",
-            retrievedItemId);
-        lookupItemHoldingsInstanceByItemId(retrievedItemId, okapiHeaders, context)
-            .onComplete(inventoryRes -> {
-          if(inventoryRes.failed()) {
-            logger.error("Unable to do inventory lookup: {}", inventoryRes.cause().getMessage());
-            promise.fail(inventoryRes.cause());
-          } else {
-            try {
-              logger.info("Attempting to populate copied items with inventory lookup for item id {}",
-                  retrievedItemId);
-              populateReserveCopiedItemFromJson(reserve, inventoryRes.result());
-              promise.complete(inventoryRes.result());
-            } catch(Exception e) {
-              promise.fail(e);
-            }
-          }
-        });
-      }
+      return Future.failedFuture("Must provide item id or item barcode to populate copied items");
+    }
+    return itemIdFuture.compose(itemIdRes -> {
```

Comment:
Moving all the code for the itemIdFuture into a
`private static Future<String> getItemId(Reserve reserve, Map<String, String> okapiHeaders, Context context)`
method might be more readable.

Reply:
Agreed.